state/action.go

```diff
@@ -385,6 +392,19 @@ func (a *action) removeAndLog(finalStatus ActionStatus, results map[string]inter
 func (a *action) removeAndLogBuildTxn(finalStatus ActionStatus, results map[string]interface{}, message string,
 	m *Model, parentOperation Operation, completedTime time.Time) jujutxn.TransactionSource {
 	return func(attempt int) ([]txn.Op, error) {
+		// If the action is already finished, return an error.
+		if a.Status() == finalStatus {
+			return nil, errors.NewAlreadyExists(nil, fmt.Sprintf("action %v already has status %q", a.Id(), finalStatus))
```

Comment:
I wanted to return an actual error here - setting the status to completed updates the "completed" time of the task. If you try to do that again a second time, what should the completed time be? Once set to completed once, I think it should be an error to do it again. The PR includes a new check in the unit agent to handle this error. ie a caller needs to opt in to knowing about the error rather than the second update being silently accepted.